Xamarin.Forms.Platform.Android/CarouselView/CarouselViewRenderer.cs

```diff
@@ -0,0 +1,607 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Android.Support.V4.View;
+using System.ComponentModel;
+
+using AViews = Android.Views;
+using System.Collections.Specialized;
+using System.Collections.Generic;
+using Com.Android.DeskClock;
+using Com.ViewPagerIndicator;
+
+/*
+The MIT License(MIT)
+
+Copyright(c) 2017 Alexander Reyes(alexrainman1975 @gmail.com)
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software
+and associated documentation files (the "Software"), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial
+portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+IN THE SOFTWARE.
+ */
+
+/*
+ * Save state in Android:
+ * 
+ * It is not possible in Xamarin.Forms.
+ * Everytime you create a view in Forms, its Id and each widget Id is generated when the native view is rendered,
+ * so its not possible to restore state from a SparseArray.
+ * 
+ * Workaround:
+ * 
+ * Use two way bindings to your ViewModel, so for example when a value is entered in a Text field,
+ * it will be saved to ViewModel and when the view is destroyed and recreated by ViewPager, its state will be restored.
+ * 
+ */
+
+namespace Xamarin.Forms.Platform.Android
+{
+	/// <summary>
+	/// CarouselView Renderer
+	/// </summary>
+	public class CarouselViewRenderer : ViewRenderer<CarouselView, AViews.View>
+	{
+		AViews.View nativeView;
+		ViewPager viewPager;
+        CirclePageIndicator indicators;
+		bool _disposed;
+
+		double ElementWidth;
+		//double ElementHeight;
+
+		protected override void OnElementChanged(ElementChangedEventArgs<CarouselView> e)
+		{
+			base.OnElementChanged(e);
+
+			if (Control == null)
+			{
+				// Instantiate the native control and assign it to the Control property with
+				// the SetNativeControl method (called when Height BP changes)
+			}
+
+			if (e.OldElement != null)
+			{
+				// Unsubscribe from event handlers and cleanup any resources
+
+				if (viewPager != null)
+				{
+					viewPager.PageSelected -= ViewPager_PageSelected;
+					viewPager.PageScrollStateChanged -= ViewPager_PageScrollStateChanged;
+				}
+
+				if (Element != null)
+				{
+					Element.SizeChanged -= Element_SizeChanged;
```

Comment:
Why not use e.OldElement instead of Element?
In this moments, Element  == e.NewElement